### ISSUE TITLE
Advancing paused time across tasks doesn't work

### ISSUE THREAD

**Issue by roblabla** [tokio-rs/tokio#3709]
**Version**: tokio v1.5.0

**Platform**: Linux 5.10.27

**Description**

Advancing the time manually (after calling `tokio::time::pause`) won't wake up `tokio::time::sleep_until` futures that happened in other spawned tasks.

I tried this code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bf1f15551f6936643cc76638e1138ab9

I expected the `worker` to wake up either during the `advance` future, or by the `yield_now` future under it. Instead, the worker never woke up from its sleep. Am I missing something?

This, coupled with #3108, makes it *really hard* to write tests that have to reason around deadlines.

**Comment by Darksonn** [comment]
This seems like it expects too much of the timer. We guarantee that the `assert_eq!(*counter.lock(), 1)` happens after the 500ms mark, and we also guarantee that `*counter += 1` happens after the 500ms mark, but I don't see any reason to expect any guarantee whatsoever regarding the ordering between those two.

**Comment by roblabla** [comment]
So, I changed the test a bit to yield in a loop, and it appears to (reliably) require 61 yields before it starts advancing the timeout task. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=63c27d8cde3d272e22704a1b1c2690d2. So I guess what's going on is that the task scheduler from tokio is deciding that my test task is more important to run than the timeout task, or something.

> This seems like it expects too much of the timer.

I mean, the timer is in a mode that's specifically made for writing tests. My expectation was that the future returned by the `advance` function would yield until all lapsed timers were woken at least once. That'd make it a useful primitive to test timers. That's not documented, but I don't think it's an unfair expectation, given the `advance` function is async! If it's not going to cause wake ups, it might as well be a sync function.

Anyways I ended up fixing it by simply not using separate tasks at all, and using `tokio::select` to swap between tasks, which works reliably and the way I expect.
